setup.cfg

```diff
@@ -100,6 +100,7 @@ install_requires =
     funcsigs>=1.0.0, <2.0.0
     graphviz>=0.12
     gunicorn>=19.5.0, <20.0
+    importlib_metadata~=1.7 # We could work with 3.1, but argparse needs <2
```

Comment:
Did you meant `argcomplete` perhaps? I note that `argcomplete` has been [updated to support `importlib_metadata` 3.x](https://github.com/kislyuk/argcomplete/blob/develop/Changes.rst) (and otherwise only has bugfixes, looks like a desireable upgrade).

And if 3.1 works, please widen the pin to `>=1.7,<4`; the `argcomplete~=1.10` pin above *already lets you use `importlib_metadata` 3.1* if you have the newer `argcomplete` installed.

Also, Python 3.8 doesn't need `importlib_metadata` anymore, and `argcomplete` correctly won't depend on it in 3.8 and up. Please do so here too:

```
     importlib_metadata>=1.7,<4;python_version<="3.7"
```

and make the imports conditional:

```python
try:
    # 3.8 and up
    from importlib import metadata as importlib_metadata
except ImportError:
    # use the backport
    import importlib_metadata
```

If you update the `argcomplete` dependency too, then you'd not even have to set the `>=1.7` lower bound, `~=3.1` would do.

Reply:
Whops, yes I did mean argcomplete. Though I also found virtualenv (dep of pylint) was also pinning this, so it may not help us yet.

This line in the readme https://github.com/python/importlib_metadata/ :

> As of Python 3.8, this functionality has been added to the Python standard library. This package supplies backports of that functionality including improvements added to subsequent Python versions

Makes me wonder if/when we should continue to use it anyway on Py 3.8?

Reply:
> Whops, yes I did mean argcomplete. Though I also found virtualenv (dep of pylint) was also pinning this, so it may not help us yet.

Ah though virtualenv has been updated too.

Reply:
> Makes me wonder if/when we should continue to use it anyway on Py 3.8?

Your not using the 3.9 / importlib_metadata 1.6 feature: `Added module and attr attributes to EntryPoint`; everything else is bug fixing or perf improvements, which all have been backported to 3.8.

Even if you _were_ using those extra attributes, using `python_version<="3.8"` (and inverting the import guard) would be a good idea to keep the dependencies flexible and avoid limiting what packages people can use with Airflow.

I don't see where `virtualenv` depends on importlib_metadata? Nor does pylint depend on virtualenv. Might be in older versions. `pre-commit` depends on virtualenv though. And jsonlint depends on `importlib_metadata` but sets no pin.

Reply:
God in getting my dist names all in a mix today. I did mean pre-commit -> virtualenv. I'll have to check what version of virtualenv pre-commit specifies.

I'll open a new PR to add the guard as you suggest.

Reply:
I've already opened a PR :-) See #12703